testing.md

```diff
@@ -0,0 +1,71 @@
+# Testing
+
+## Why write tests?
+
+We write automated tests to allow us to quickly demonstrate that our software meeting business requirements. They're one of a few tools we use to increase our confidence in our work prior to deployment.
+
+## When do I write tests?
+
+### Test necessity
+_"If this test fails, what business requirement is compromised?"_
+
+This is a good question to ask yourself before writing tests. If the answer is _"None"_, you shouldn't be writing it. If the answer is _"I don't know"_, it's possible the test you're writing has no value.
+
+### Net costs
+It's also worth considering the net cost of tests before you write them. For example, if the business impact of a regression is small and unlikely enough, it could be that you're probably incurring a net cost to the business by writing a test that needs to be **1.** understood, **2.** run as part of your CI and **3.** maintained.
+
+### Keep the system in mind
+Lastly, when making decisions about the scope of your testing, keep the system in mind that you're working on. For example, a single failure on a data-sensitive area of the API is potentially going to affect hundreds of customers and could expose masses of customer data. Whereas a single failure in React component is only going to affect a single session at once, and probably won't be a showstopper.
+
+## Types of Testing
+There are three major categories of testing that might be appropriate to use in your development.
+
+### Unit Testing
+[Unit testing](https://en.wikipedia.org/wiki/Unit_testing) involves testing code at a very granular level, usually testing a single function and everything that goes on within it, stubbing calls to other functions or modules to ensure we can accurately check the functionality of the unit. Taking making a cake as an analogy, you can think of unit testing as checking the quality of the individual ingredients.
+
+Unit tests are particularly useful for things like:
+ - Proving the functionality of complex algorithmic logic.
+ - Alerting developers to potential regressions during refactors in a "yes, I really did mean to do that there" kind of fashion.
+ - Providing 100% thorough exercising of business critical code. E.g. Ensure a card number validator does not block valid cards.
+ - Testing potential edge cases for functions. E.g. "What happens when I throw garbage into this?"
+
+However, unit tests have some drawbacks:
+ - Expensive to write
+ - Expensive to maintain
+ - Only accessible to developers
+ - Hard to determine a business purpose from. E.g. "I can see this code is doing this and that, but what does that actually _mean_ in product terms?"
```

Comment:
I included all of these points because people tend to write either unit or system tests at Hx. By laying down all the pros and cons of them from an overall quality assurance perspective, I'm hoping people are able to make informed choices themselves about what's best for their code. I agree that the last point is more the point of unit testing than an inherent disadvantage though.